Extension/src/LanguageServer/configurations.ts

```diff
@@ -2050,13 +2067,29 @@ export class CppProperties {
                     if (curOffset >= compilerPathStart && curOffset <= compilerPathEnd) {
                         continue;
                     }
-                    let message: string;
+                    let message: string = "";
                     if (!pathExists) {
                         if (curOffset >= forcedIncludeStart && curOffset <= forcedeIncludeEnd
                                 && !path.isAbsolute(resolvedPath)) {
                             continue; // Skip the error, because it could be resolved recursively.
                         }
-                        message = localize('cannot.find2', "Cannot find \"{0}\".", resolvedPath);
+                        // If there are incorrect paths and the first match includes "env:", split the string and check for environment variable.
+                        if (incorrectExpandedPaths.length > 0 && configMatches[0].includes("env:")) {
+                            const splitString = configMatches[0].split("env:");
+                            if (splitString[1]) {
+                                const matchResult = splitString[1].match(/\w+/);
+                                if (matchResult) {
+                                    const envVar = matchResult[0];
+                                    message = localize('cannot.find2', "Cannot find {0} in {1} environment variable.", incorrectExpandedPaths.map(s => `"${s}"`).join(', '), envVar);
+                                }
+                            }
+                        // If there are incorrect paths but no "env:" in the first match, generate a message indicating the paths cannot be found.
+                        } else if (incorrectExpandedPaths.length > 0) {
+                            message = localize('cannot.find2', "Cannot find {0}.", incorrectExpandedPaths.map(s => `"${s}"`).join(', '));
+                        } else {
+                            message = localize('cannot.find2', "Cannot find \"{0}\".", resolvedPath);
```

Comment:
You can't have two different messages with the same localize id.  Localizers mess up the quotes anyway, so you should do this instead:

```ts
else {
  let badPath = "";
  if (incorrectExpandedPaths.length > 0) {
    badPath = incorrectExpandedPaths.map(s => `"${s}"`).join(', ');
  } else {
    badPath = `"${resolvedPath}"`;
  }
  message = localize('cannot.find2', "Cannot find {0}", badPath);
}
```

Or you could push `resolvedPath` into `incorrectExpandedPaths` and not have a conditional here.  I haven't reviewed the whole thing, but am wondering why you have two different variables tracking essentially the same thing.  If possible, push all invalid paths into the array so you only have to work with one variable.